public/app/plugins/datasource/tempo/dataquery.cue

```diff
@@ -44,6 +44,8 @@ composableKinds: DataQuery: {
 					maxDuration?: string
 					// Filters to be included in a PromQL query to select data for the service graph. Example: {client="app",service="app"}
 					serviceMapQuery?: string
+					// Use service.namespace in addition to service.name to uniquely identify a service. 
+					serviceMapIncludeNamespace?: bool
```

Comment:
Would it make sense to instead make this something like: `serviceMapGroupByLabels?: string[]` and make it more flexible? In case different labels would be used somewhere or different groupings? If these things can be configured in the agent?

Reply:
> Would it make sense to instead make this something like: serviceMapGroupByLabels?: string[] and make it more flexible? In case different labels would be used somewhere or different groupings? If these things can be configured in the metric generator?

I pondered this! But there's no alternative use cases right now and no idea how to represent more than 1 additional dimension. So thought it best to keep it simple. Could refactor later when and if we find a use case though

Reply:
> no idea how to represent more than 1 additional dimension in the node panel

Not sure I get this. In essence we are making multiple groups where each group is actually a separate node right? More groups (labels in the `sum by`) will just make the metrics more granular. If I get the code the problem is that first we need to specify what are the labels to group by, then understand which ones are server which client and then later on when creating the data frame we need to create new nodeIDs from multiple labels. All this right now is sort of hardcoded (with `serviceMapIncludeNamespace` being either on or off).

So I agree this is the easiest now, but wonder if we could make it flexible while still pretty simple by using https://prometheus.io/docs/prometheus/latest/querying/functions/#label_join or https://prometheus.io/docs/prometheus/latest/querying/functions/#label_replace. Because what you are essentially doing is saying `serverID = server + server_namespace`. If we could just relabel the response right away only thing we would need to modify is the query and add the relabel there and we would not have to change any other code. That way this could be more like `serviceMapServeIdLabels?: string[]` and `serviceMapClientIdLabels?: string[]` which would default to `client` and `server` labels in your case `serviceMapClientIdLabels: ['client', 'client_namespace']` and `serviceMapClientIdLabels: ['server', 'server_namespace']`.

Would that make sense? That flexibility probably isn't needed right now but again I assume this would mean we don't have to change the dataTransform logic and there would be less hardcoding of labels. But yeah if there are some complications I don't know about I guess we could go with a little hardcoding for now.

Reply:
Ahh, sorry, let me elaborate: the data pipeline changes are easy, even modifying the transform is not a big deal. By "no idea how to represent" I mean UX. If there are more dimensions, they need to be readable in the node graph visualization. Right now I just put the namespace in the subtitle, which seems obvious and good enough. But what to do if there are multiple extra dimensions? Then probably a more complex UX solution is needed, and I'm not sure it's worthwhile to grapple with it until we have a solid use case and more data to inform the design.

Another point: `service.name`, `service.namespace` and `service.id` are semnatically intended as canonical identifers of a service. IMHO it makes the most sense to add all data required to uniquely identify a service into `service.name` and `service.namespace` anyway instead of using other labels, this will make it work with other tooling (eg, app o11y plugin) without extra configuration

Does that makes sense? 
If you still think we should make this more generic, we'll work on it